### DISCUSSION TITLE
"Coverage 100%" badge is misleading

### DISCUSSION THREAD

**Discussion by mrazauskas** [wellwelwel/poku#613] [General]
While trying to migrate a larger code base to Poku, I bump into an issue with async `afterEach()` and async `test()`. Moving code from `afterEach()` to the end of `test()` makes all work as expected. (I will create reproduction later.)

Out of curiosity I looked at: [`src/modules/helpers/test.ts`](https://github.com/wellwelwel/poku/blob/main/src/modules/helpers/test.ts). If I get it right, the `test()` helper logic lives here. Unfortunately the promise handling logic is wrapped with `c8 ignore` comments. Hm.. Might be I miss something, but at first glance it does not like these lines are not reachable.

"Coverage 100%" badge looks beautiful, but now for me it feels misleading. The high coverage rate was one of the aspects why I decided to migrate to Poku. Please consider updating the badge with actual coverage information.

**Comment by wellwelwel** [comment]
Hi, @mrazauskas.

The only mention for this is in [**Contributing.md**](https://github.com/wellwelwel/poku/blob/main/CONTRIBUTING.md#%EF%B8%8F-coverage):

https://github.com/wellwelwel/poku/blob/2b1eff045f1d7c2804b26fcb18120d3eb7042297/CONTRIBUTING.md#L116-L118

> "Don't be intimidated by high coverage, methods that vary according to platform, platform versions, _OS_ and processes _(`process.exit`, `process.once`, etc.)_ aren't tested against the coverage rate 🙋🏻‍♂️"

---

I use two specific comments related to a problem between **c8** and **tsx**:

```js
/* c8 ignore next */ // ?
```

- For some reason, the first line of some files and the last export always appear as zero coverage, including when the line is a comment.

```js
/* c8 ignore next */ // Types
```

- When using overloads or importing types (`ts`), **tsx** build removes it before running, causing coverage mistakes.

---

Except by these two, all other comments should fit on `process`, `Windows`, previous versions from **Node.js** or vary for **Bun** and **Deno** platforms.

That line you mentioned, there's really no point in ignoring it in the coverage. I apologize for any misunderstanding and would like to hear your opinion on these exceptions I mentioned. 100% coverage is indeed **Poku**'s goal.

---

> Being honest, I loved the nudge and will revisit all the **c8** comments, thank you for that.

Also, I'm considering migrating from **c8** to **monocart-coverage-reports**.

**Comment by mrazauskas** [comment]
All is fine. This is only a badge (;

By the way, I was thinking to recommend looking at `monocart-coverage-reports`. It is really good in filtering out type only or empty lines. And the native `v8` / `codecov` coverage is their strong side.

**Comment by mrazauskas** [comment]
By the way, those `c8 ignore` comments are shipped to `npm` and increase the install size without being anyhow useful. Just another argument (;

**Comment by wellwelwel** [comment]
> By the way, those `c8 ignore` comments are shipped to `npm` and increase the install size without being anyhow useful. Just another argument (;

Yes, it bothers me a lot.

I tried a test locally with `monocart-coverage-reports` and only 8 comments remained 🎉
Most due to `process`, for example:

```ts
/* c8 ignore next 3 */ // Process
export const onSigint = () => {
  process.stdout.write('\u001B[?25h');
};

/* c8 ignore next */ // Process
process.once('SIGINT', onSigint);
```

---

I'll try to investigate some way of being able to test specific scenarios with `process`.

**Comment by wellwelwel** [comment]
# 👥 For community

> [!NOTE]
>
> If you don't agree with any of these not covered topics, please feel free to comment. I'm totally open to discussions and opinions.

### Different behaviors due to platform versions

- Compatibility with `c8` starts with [**Node.js** `v18`](https://github.com/bcoe/c8/blob/ff146b4dde004c62651b57c33cedd8353c94c423/package.json#L67).
- Compatibility with `monocart-coverage-reports` starts with [**Node.js** `v18`](https://github.com/cenfun/monocart-coverage-reports/issues/60).

### Different behaviors due to **Deno** and **Bun** platforms

- At the moment, there is no way to generate the coverage report for them using `c8` or `monocart-coverage-reports`.

### process-based

> E.g., `process.exit`, `process.once`, etc.

- I intend to do more research.

### Different behaviors due to _OS_

> **Windows**, specifically.

- I'm planning to remove this one, since we can use **Windows** matrix to ensure this.

---

The choice not to consider these topics in the coverage comes from the fact that it isn't possible, not that they are ignored. For example, even if exhaustive tests are created for **Node.js** `v8`, there is no way to generate the coverage report for these tests.

> Similarly, there are specific tests for **Bun** and **Deno** that don't generate coverage reports.

In disagreements, this issue can be reopened at any time 🤝
 


---

## Related

- https://github.com/wellwelwel/poku/pull/585

**Comment by wellwelwel** [comment]
> _[...]_ The high coverage rate was one of the aspects why I decided to migrate to Poku
> _[...]_ but now for me it feels misleading

Less formal, I don't want anyone to feel that way when using **Poku**, which is why I'm taking it so seriously 💙

**Comment by mrazauskas** [comment]
I appreciate the effort. Thank you.

**Reply by wellwelwel** [reply]
@mrazauskas, thank you for your support. After #612, I feel more comfortable continuing to develop new features for **Poku**.

Also, `monocart-coverage-reports` reported points that `c8` pointed out as fully covered.

**Reply by mrazauskas** [reply]
Looks good. I just wonder why you decided to keep `c8` instead of using the `mcr` binary directly?

**Reply by wellwelwel** [reply]
> I just wonder why you decided to keep `c8` instead of using the `mcr` binary directly?

I didn't want to, but `monocart-coverage-reports` fails when the failure tests come up 😕

In combination with `c8`, it worked normally.

**Reply by mrazauskas** [reply]
Interesting. Just open an issue in their repo. They respond quickly and are happy to help.

**Reply by wellwelwel** [reply]
> Interesting. Just open an issue in their repo. They respond quickly and are happy to help.

Thanks, I'll try to isolate a minimal reproduction 🙋🏻‍♂️
